src/vos/vos_iterator.c

```diff
@@ -720,6 +720,19 @@ vos_iterate_internal(vos_iter_param_t *param, vos_iter_type_t type,
 				D_GOTO(out, rc);
 
 			reset_anchors(iter_ent.ie_child_type, anchors);
+
+			/** The child iterator may yield during iteration.
+			 *  It isn't necessary to reprobe the parent in order
+			 *  to continue iterating on the child because it
+			 *  would remain in the same location in memory.
+			 *  However, it is necessary to reprobe this iterator
+			 *  before continuing.
+			 */
+			if (need_reprobe(type, anchors)) {
+				D_ASSERT(!daos_anchor_is_zero(anchor) &&
+					 !daos_anchor_is_eof(anchor));
+				goto probe;
```

Comment:
To make code clear, I think It's better to move the following need_reprobe() call (before vos_iter_next()) into the if (post_cb), since it's only for current level reprobe caused by post_cb() now. (It was used for handling both current reprobe and parent reprobe before the post_cb code was added).

Reply:
Ok, I'll update if I need a refresh for any reason.  Otherwise, will leave to another patch.